ci: Build python from source in "lint" task

hebasto opened the pull request:
This PR:
- is an alternative to bitcoin/bitcoin#26581 and bitcoin/bitcoin#26637
- closes bitcoin/bitcoin#26548

Key advantages of this PR over others:
- it uses pyenv's `python-build` [standalone](https://github.com/pyenv/pyenv/tree/master/plugins/python-build#using-python-build-standalone)
- requires no additional computational resources

Note for testing. The lint task must success regardless of whether the `python_cache` is populated or invalidated.

fanquake commented:
>  it uses pyenv's python-build [standalone](https://github.com/pyenv/pyenv/tree/master/plugins/python-build#using-python-build-standalone); no other tools required

As far as I can tell, `pyenv install` is just [a wrapper around](https://github.com/pyenv/pyenv/blob/master/COMMANDS.md#pyenv-install) `python-build`. You still need the same tools (compiler / linker / all the python build dependencies) to compile Python regardless of how you compile it. I guess the difference here is just not fully installing pyenv? 

> requires no additional computational resources

GIven that `pyenv install` and `python-build` are the same thing, why is there a difference in computational resources required?

MarcoFalke commented on ci/lint/04_install.sh:
```diff
@@ -7,14 +7,22 @@
+    cd pyenv/plugins/python-build || exit 1
+    ./install.sh
+  )
+  ${CI_RETRY_EXE} apt-get install -y build-essential clang libbz2-dev liblzma-dev libncursesw5-dev libreadline-dev libssl-dev libsqlite3-dev zlib1g-dev
```
might be nice to add a comment where the deps are copy-pasted from

hebasto replied:
Added.

hebasto commented:
> > it uses pyenv's python-build [standalone](https://github.com/pyenv/pyenv/tree/master/plugins/python-build#using-python-build-standalone); no other tools required
> 
> As far as I can tell, `pyenv install` is just [a wrapper around](https://github.com/pyenv/pyenv/blob/master/COMMANDS.md#pyenv-install) `python-build`. You still need the same tools (compiler / linker / all the python build dependencies) to compile Python regardless of how you compile it. I guess the difference here is just not fully installing pyenv?

Correct. Dropped "no other tools required" from the PR description.

> > requires no additional computational resources
> 
> GIven that `pyenv install` and `python-build` are the same thing, why is there a difference in computational resources required?

I reckon the lower memory usage is due to using clang instead of gcc.

MarcoFalke commented on .cirrus.yml:
```diff
@@ -64,14 +64,18 @@ compute_credits_template: &CREDITS_TEMPLATE
+  python_cache:
+    folder: "/tmp/python"
+    fingerprint_script: cat .python-version /etc/os-release
+    populate_script: rm -rf /tmp/python
```
Is this needed?

hebasto replied:
> Is this needed?

No, it is not. [Dropped](https://github.com/bitcoin/bitcoin/pull/26716#issuecomment-1382713392).

hebasto commented:
Updated 3f602a628bdb8007c2ebebe58c10f42db787679f -> f99260fdd02453649222bfe4d38d0655ebdbc42c ([pr26716.02](https://github.com/hebasto/bitcoin/commits/pr26716.02) -> [pr26716.03](https://github.com/hebasto/bitcoin/commits/pr26716.03), [diff](https://github.com/hebasto/bitcoin/compare/pr26716.02..pr26716.03)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/26716#discussion_r1064577306)

MarcoFalke commented on ci/lint/04_install.sh:
```diff
@@ -7,14 +7,23 @@
+    ./install.sh
+  )
+  # For dependencies see https://github.com/pyenv/pyenv/wiki#suggested-build-environment
+  ${CI_RETRY_EXE} apt-get install -y build-essential clang libbz2-dev liblzma-dev libncursesw5-dev libreadline-dev libssl-dev libsqlite3-dev zlib1g-dev
```
It doesn't look like this list is copied from the above link, why?

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/26716#issuecomment-1383865336).

hebasto commented:
Updated f99260fdd02453649222bfe4d38d0655ebdbc42c -> d418803a6a60bfb0d51ed5deb658ffd7639f31d5 ([pr26716.03](https://github.com/hebasto/bitcoin/commits/pr26716.03) -> [pr26716.05](https://github.com/hebasto/bitcoin/commits/pr26716.0), [diff](https://github.com/hebasto/bitcoin/compare/pr26716.03..pr26716.05)):

- all documented dependencies being added explicitly, addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/26716#discussion_r1071051411)

hebasto commented:
CI tasks with the python cache being:
- populated -- https://cirrus-ci.com/task/4912581312249856
- invalidated -- https://cirrus-ci.com/task/6641128783478784

hebasto commented on ci/lint/04_install.sh:
```diff
@@ -18,22 +18,15 @@ if [ ! -d "${PYTHON_PATH}/bin" ]; then
+    libbz2-dev libreadline-dev libsqlite3-dev curl clang llvm \
     libncursesw5-dev xz-utils tk-dev libxml2-dev libxmlsec1-dev libffi-dev liblzma-dev
-  python-build "$(cat "${BASE_ROOT_DIR}/.python-version")" "${PYTHON_PATH}"
+  # Using clang to workaround https://github.com/pyenv/pyenv/issues/2359.
```
> How/why does using Clang solve the segmentation fault issue?

I have no answers, unfortunately.

> Also seems unecessary to put this in the second commit, instead of the first.

There were no problems with gcc on bionic.

fanquake replied:
> I have no answers, unfortunately.

Well linking to this thread, that doesn't explain anything, and just has a comment with a link to a comment in a different thread, where someone claimed that using Clang magically solved their problem, is more confusing than useful.

MarcoFalke replied:
Presumably a gcc compiler bug?

hebasto replied:
> Presumably a gcc compiler bug?

Perhaps. But do we really want to dig into gcc bugtracker for the sake of building an old python?

MarcoFalke replied:
We'll need clang either way, because it uses less memory

hebasto replied:
Does this comment look good
```
# Using clang for less memory footprint and to workaround https://github.com/pyenv/pyenv/issues/2359.
```
?

hebasto replied:
> Also seems unecessary to put this in the second commit, instead of the first.

Done.

> I don't think there is any value in linking to [pyenv/pyenv#2359](https://github.com/pyenv/pyenv/issues/2359).

Dropped.

hebasto committed:
ci: Use pyenv's `python-build` to install Python in lint task

hebasto committed:
ci: Bump lint task image to Ubuntu Jammy

fanquake reviewed (approved):
ACK 123043e99cf3aab9eef7e381b133477b518ac4d0

MarcoFalke commented on ci/lint/04_install.sh:
```diff
@@ -7,14 +7,25 @@
-  ${CI_RETRY_EXE} apt-get update
-  ${CI_RETRY_EXE} apt-get install -y --reinstall git
-)
+${CI_RETRY_EXE} apt-get install -y curl git gawk jq xz-utils
```
Why add `xz-utils` here when it is only needed when compiling python?

hebasto replied:
It is required here: https://github.com/bitcoin/bitcoin/blob/01ec5308bf616740c804247043046b23b483df5c/ci/lint/04_install.sh#L37

jamesob commented:
This change points to the wastefulness of our current CI setup as alluded to in [the Cirrus docs](https://cirrus-ci.org/guide/docker-builder-vm/):

![image](https://user-images.githubusercontent.com/73197/212981230-97300355-0983-4444-8fdb-7ad85ee742de.png)

On every CI run, we're doing the same apt installs and Python setup. We should probably just specify a `Dockerfile`, the resulting image of which can be cached on the Cirrus side. I think that would also simplify CI on our end and make it easier to run locally. 

If there's interest I can create an issue for this.

MarcoFalke commented:
Yeah, I did this because I want to be able to run the stuff easily locally: https://github.com/bitcoin/bitcoin/tree/master/ci#running-a-stage-locally

I don't know enough about docker to figure out if there is a more efficient but equally simply way to do this.


Overall I am thinking that it shouldn't matter if we download the packages or download a pod layer with the packages installed. (Modulo a slow mirror, which should ideally be solved by using a fast mirror, but there doesn't seem to be an easy way to do this? Also modulo the time it takes to run the install script)
